ILSpy.ReadyToRun/ReadyToRunLanguage.cs

```diff
@@ -0,0 +1,170 @@
+// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
+// to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or
+// substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Runtime.CompilerServices;
+using Iced.Intel;
+using ICSharpCode.Decompiler;
+using ICSharpCode.Decompiler.Metadata;
+using ICSharpCode.Decompiler.Solution;
+using ICSharpCode.Decompiler.TypeSystem;
+using ILCompiler.Reflection.ReadyToRun;
+
+namespace ICSharpCode.ILSpy
+{
+	[Export(typeof(Language))]
+	class ReadyToRunLanguage : Language
+	{
+		private static readonly ConditionalWeakTable<PEFile, R2RReader> r2rReaders = new ConditionalWeakTable<PEFile, R2RReader>();
+		public override string Name => "ReadyToRun";
+
+		public override string FileExtension {
+			get { return ".asm"; }
+		}
+
+		private R2RReader GetReader(LoadedAssembly assembly, PEFile module)
+		{
+			R2RReader result;
+			lock (r2rReaders) {
+				if (!r2rReaders.TryGetValue(module, out result)) {
+					// TODO: avoid eager parsing 
+					result = new R2RReader(new R2RAssemblyResolver(assembly), module.Metadata, module.Reader, module.FileName);
+					r2rReaders.Add(module, result);
+				}
+			}
+			return result;
+		}
+
+		public override ProjectId DecompileAssembly(LoadedAssembly assembly, ITextOutput output, DecompilationOptions options)
+		{
+			PEFile module = assembly.GetPEFileOrNull();
+			R2RReader reader = GetReader(assembly, module);
+
+			WriteCommentLine(output, "TODO - display ready to run information");
+			// TODO: display other header information
+			foreach (var method in reader.R2RMethods) {
+				WriteCommentLine(output, method.SignatureString);
+			}
+
+			return base.DecompileAssembly(assembly, output, options);
+		}
+
+		public override void DecompileMethod(IMethod method, ITextOutput output, DecompilationOptions options)
+		{
+			PEFile module = method.ParentModule.PEFile;
+			R2RReader reader = GetReader(module.GetLoadedAssembly(), module);
+			int bitness = -1;
+			if (reader.Machine == Machine.Amd64) {
+				bitness = 64;
+			} else if (reader.Machine == Machine.I386) {
+				bitness = 32;
+			}
+			else {
+				// TODO: Architecture other than x86/amd64
+				throw new NotImplementedException("");
```

Comment:
for the MVP version I would prefer to simply display "; Architecture xxxx is currently not supported." in the output rather than having ILSpy crash.

Reply:
On a related note: please do not crash in case an assembly does not contain any R2R info at all. You could simply display  "; This file does not contain a ReadyToRun assembly." or similar.